service/history/workflowTaskHandler.go

```diff
@@ -229,19 +242,87 @@ func (handler *workflowTaskHandlerImpl) handleCommandScheduleActivity(
 	)
 	if err != nil || failWorkflow {
 		handler.stopProcessing = true
-		return err
+		return nil, err
 	}
 
 	enums.SetDefaultTaskQueueKind(&attr.GetTaskQueue().Kind)
 
-	_, _, err = handler.mutableState.AddActivityTaskScheduledEvent(handler.workflowTaskCompletedID, attr)
+	localDispatchActivity := false
+	namespace := handler.mutableState.GetNamespaceEntry().Name().String()
+	if attr.RequestStart &&
+		targetNamespaceID == namespaceID && // TODO: shall we return error, other check?
+		handler.config.EnableActivityLocalDispatch(namespace) {
+		localDispatchActivity = true
+	}
+
+	event, ai, err := handler.mutableState.AddActivityTaskScheduledEvent(
+		handler.workflowTaskCompletedID,
+		attr,
+		localDispatchActivity,
+	)
 	if err != nil {
 		if _, ok := err.(*serviceerror.InvalidArgument); ok {
-			return handler.failCommand(enumspb.WORKFLOW_TASK_FAILED_CAUSE_SCHEDULE_ACTIVITY_DUPLICATE_ID, err)
+			return nil, handler.failCommand(enumspb.WORKFLOW_TASK_FAILED_CAUSE_SCHEDULE_ACTIVITY_DUPLICATE_ID, err)
 		}
-		return err
+		return nil, err
 	}
-	return nil
+
+	if !localDispatchActivity {
+		return nil, nil
+	}
+
+	if _, err := handler.mutableState.AddActivityTaskStartedEvent(
```

Comment:
does server control how many activities are allowed for local dispatch? or all of them as SDK asked?

Reply:
Synced offline. Currently there's no obvious reason for server to reject the request (even if the activity is scheduled cross namespace or on a different task queue). The workflow mutable state update transaction and history have the same size.

If worker set `request_start` to true, then it should guarantee it is able to run that activity.